# Current Topics in Bioinformatics workshops

| Audience | Computational Skills | Duration |
:----------|:----------|:----------|
| Biologists | Beginner or intermediate R and/or beginner bash | 2-3 hour workshops |

This repository has teaching materials for **2-3 hour**, hands-on workshops covering a variety of topics related to bioinformatics data analysis. The workshops will lead participants through performing different types of analyses using R/RStudio or Linux. 

Some workshops will require a working knowledge of R or completion of the [Introduction to R workshop](https://hbctraining.github.io/Training-modules/IntroR). Other workshops will require a working knowledge of the bash scripting language or completion of the [Introduction to Unix workshop](https://hbctraining.github.io/Training-modules/Intro_shell/). 

See our current workshop schedule on our [training website](http://bioinformatics.sph.harvard.edu/training/#current-topics-in-bioinformatics---spring-2019).

** **NOTE: Detailed information** and **preparation instructions** for each of the workshops can be found by clicking on the workshop links in the table below.

> Workshop introduction slides are available [here](https://github.com/hbctraining/Training-modules/raw/master/Intro_current_topics.pdf).

### Current Workshops

| Lessons        | Prerequisites |
|:---------------|:-------------:|
| [Planning a bulk RNA-seq analysis: Part I](https://hbctraining.github.io/Training-modules/planning_successful_rnaseq#part-i) | None |
| [Planning a bulk RNA-seq analysis: Part II](https://hbctraining.github.io/Training-modules/planning_successful_rnaseq#part-ii) | None |
| [Make your (RNA-seq) data analysis reproducible](https://hbctraining.github.io/Training-modules/reproducible_analyses/reproducible_data_analysis.md)- *Taught by [Julie Goldman](https://scholar.harvard.edu/julie_goldman) from Countway Library* | None |
| Improving your (RNA-seq) data analysis using version control (Git) | None |

### All topics

| Lessons        | Prerequisites |
|:---------------|:-------------:|
| [Introduction to R](https://hbctraining.github.io/Training-modules/IntroR) | None |
| [Introduction to the tidyverse data science packages and visualizations with ggplot2](https://hbctraining.github.io/Training-modules/Tidyverse_ggplot2) | Beginner R or [IntroR workshop](https://hbctraining.github.io/Training-modules/IntroR) |
| [Gene annotations and functional analysis of gene lists](https://hbctraining.github.io/Training-modules/DGE-functional-analysis/) | Beginner R or [IntroR workshop](https://hbctraining.github.io/Training-modules/IntroR) |
| [Generating research analysis reports with RMarkdown](https://hbctraining.github.io/Training-modules/Rmarkdown/) | Beginner R or [IntroR workshop](https://hbctraining.github.io/Training-modules/IntroR) |
| [Visualization in R](https://hbctraining.github.io/Training-modules/Visualization_in_R/) | Beginner R or [IntroR workshop](https://hbctraining.github.io/Training-modules/IntroR) |
| [Functional analysis of gene lists](https://hbctraining.github.io/Training-modules/DGE-functional-analysis/) | Beginner R or [IntroR workshop](https://hbctraining.github.io/Training-modules/IntroR) |
| [Reproducible research using R (Rmarkdown: report generation)](https://hbctraining.github.io/Training-modules/Rmarkdown/) | Beginner R or [IntroR workshop](https://hbctraining.github.io/Training-modules/IntroR) |
| [Introduction to the command-line interface](https://hbctraining.github.io/Training-modules/Intro_shell/) | None |
| [Intermediate bash](https://hbctraining.github.io/Training-modules/Intermediate_shell/) | Beginner bash or [Intro to the command-line interface](https://hbctraining.github.io/Training-modules/Intro_shell/) |
| [Version control using Git and Github](https://hbctraining.github.io/Training-modules/Git-Github) | Beginner bash or [Intro to the command-line interface](https://hbctraining.github.io/Training-modules/Intro_shell/)  |
| [Accessing genomic reference and experimental sequencing data](https://hbctraining.github.io/Accessing_public_genomic_data) | Beginner bash or [Intro to the command-line interface](https://hbctraining.github.io/Training-modules/Intro_shell/)  |
| [Exploring genomic variants using GEMINI](https://hbctraining.github.io/Training-modules/Exploring_variants_with_GEMINI) | Beginner bash or [Intro to the command-line interface](https://hbctraining.github.io/Training-modules/Intro_shell/) |

***

*These materials have been developed by members of the teaching team at the [Harvard Chan Bioinformatics Core (HBC)](http://bioinformatics.sph.harvard.edu/). These are open access materials distributed under the terms of the [Creative Commons Attribution license](https://creativecommons.org/licenses/by/4.0/) (CC BY 4.0), which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited.*
