## Exploring genomic variants using GEMINI

| Audience | Computational Skills | Prerequisites | Duration |
:----------|:----------|:----------|:----------|
| Biologists | Intermediate bash | Introduction to the command-line interface | 3 hour workshop (~3 hours of trainer-led time)|


### Description

This repository has teaching materials for a **3 hour**, hands-on **Exploring genomic variants using GEMINI** workshop led at a relaxed pace. [GEMINI (GEnome MINIng)](https://github.com/arq5x/gemini) is a tool that helps turn those giant, sparse VCF variant matrices (millions of rows, thousands of columns) into a simple, accessible database. Within the database GEMINI annotates with just about everything out there. ENCODE, OMIM, dbSNPâ€¦ plus internal annotations like regions of interest, candidate genes, etc. The resulting framework supports an interactive exploration of variant information in the context of numerous third-party genomic annotations.

### Learning Objectives

* Using annotation information to filter out important variants
* Creating a GEMINI database using VCF files
* Exploring how to query the GEMINI database
* Extracting important variants based on characteristics, such as severity, type, location, etc.

> These materials are developed for a trainer-led workshop, but also amenable to self-guided learning.


### Contents

| Lessons            | Estimated Duration |
|:------------------------|:----------|
|[Exploring genomic variants using GEMINI](https://hbctraining.github.io/Intro-to-Shell/lessons/04_loops_and_scripts.html) | 75 min |

### Installation Requirements

***Mac users:***
No installation requirements.

***Windows users:***
[GitBash](https://git-scm.com/download/win)
